src/types.rs

```diff
@@ -825,6 +825,169 @@ impl SourceMap {
+    /// its target file is the source file of `right`. In other words, if `left`
+    /// maps from `transformed.js` to `minitfied.js` and `right`
+    /// maps from `minified.js` to `original_1.js`, …, `original_n.js`, then
+    /// the resulting sourcemap maps from `transformed.js`
+    /// to to `original_1.js`, …, `original_n.js`.
+    ///
+    /// The source root, sources, source contents, and names will be copied from `right`. The only information
+    /// that is used from `first` are the mappings.
+    pub fn compose(left: &Self, right: &Self) -> Self {
+        // The composition works by going through the tokens in `right` in order and adjusting
+        // them depending on the token in `left` they're "covered" by.
+        // For example:
+        // Let `l` be a token in `left` mapping `(17, 23)` to `(8, 30)` and let
+        // `r₁ : (8, 32) -> (102, 35)`, `r₂ : (8, 40) -> (102, 50)`, and
+        // `r₃ : (9, 10) -> (103, 12)` be the tokens in `right` that fall in the range of `l`.
+        // `l` offsets these tokens by `(+9, -7)`, so `r₁, … , r₃` must be offset by the same
+        // amount. Thus, the composed sourcemap will contain the tokens
+        // `c₁ : (17, 25) -> (102, 35)`, `c₂ : (17, 33) -> (102, 50)`, and
+        // `c3 : (18, 3) -> (103, 12)`.
+        // Moreover, there is a small gap between the start of the range of `l` (`(8, 30)`) and the start
+        // of the domain of `r₁` (`(8, 32)`). Therefore, we also add a token `c₀ : (8, 30) -> !` that maps
+        // the small range `(8, 30)..(8, 32)` to nothing.
+
+        // Helper struct that makes it easier to compare tokens by the start and end
+        // of the range they cover.
+        #[derive(Debug, Clone, Copy)]
+        struct Range {
+            start: (u32, u32),
+            end: (u32, u32),
+            value: RawToken,
+        }
+        let mut builder = SourceMapBuilder::new(right.file.as_deref());
+        builder.set_source_root(right.get_source_root());
+
+        // Turn `left.tokens` and `right.tokens` into vectors of ranges so we have easy access to
+        // both start and end.
+        // We want to compare `left` tokens and `right` by line/column numbers in the "middle" file.
+        // These line/column numbers are the `src_line/col` for `left` tokens and `dst_line/col` for
+        // the right tokens.
+        let mut left_tokens = left.tokens.clone();
+        left_tokens.sort_unstable_by_key(|t| (t.src_line, t.src_col));
+        let mut left_token_iter = left_tokens.iter().peekable();
+        let mut left_ranges = Vec::new();
+
+        while let Some(&t) = left_token_iter.next() {
+            let (end_line, end_col) = left_token_iter
+                .peek()
+                .map_or((u32::MAX, u32::MAX), |&&t| (t.src_line, t.src_col));
+            left_ranges.push(Range {
+                start: (t.src_line, t.src_col),
+                end: (end_line, end_col),
+                value: t,
+            });
+        }
```

Comment:
I believe you can use the [`slice::windows`](https://doc.rust-lang.org/std/primitive.slice.html#method.windows) iterator for this and chaining another `u32::MAX` end line on to the end.

Another idea is to dedupe the duplicated code here into a function that takes an accessor closure that return `src_X` or `dst_X`.

Reply:
I'll see if that ends up looking better, thanks.

Reply:
Update: I tried a version with `windows` and it has the same flaws as using `peek`, which I tried earlier. It makes comparisons between the starts and ends of ranges much more cumbersome and less readable. I believe the extra work to define and compute `Range`s is worth it.

Refactoring out the common code is definitely a good idea, though.